### PR TITLE
Add @include_dir directive: debian-style ("conf.d") config processing

### PR DESCRIPTION
**Pull request by jltallon**
Quite some refactoring of the includefile stack and filename processing logic

  Adds some extra fields to scan_context, but does not change the API
  "symbols" file should hide all functions called scanctx_* anyway

### CODE REVIEW

**Comment by jltallon** [comment]
@include_dir feature (first commit) plus documentation and a bit more robustness (second commit)

File: lib/scanctx.c [hunk 1]
```diff
@@ -92,43 +93,26 @@ const char **scanctx_cleanup(struct scan_context *ctx,
...
-  free((void *)full_file);
-
+  fp = fopen(file, "rt");
   if(fp)
   {
     ctx->streams[ctx->depth] = fp;
     ctx->files[ctx->depth] = __scanctx_add_filename(ctx, file);
     ctx->buffers[ctx->depth] = buffer;
     ++(ctx->depth);
+	*error = NULL;
```
Comment (hyperrealm) [inline lib/scanctx.c]: incorrect indentation

File: lib/scanctx.c [hunk 2]
```diff
@@ -137,6 +121,40 @@ FILE *scanctx_push_include(struct scan_context *ctx, void *buffer,
...
+
+  return full_path;
+}
+
+
+/* ------------------------------------------------------------------------- */
+
+const char *scanctx_filename(struct scan_context *ctx,const char* dirname, const char* filename)
+{
+ const char* basedir = (NULL!=dirname)? dirname : ctx->basedir;
```
Comment (hyperrealm) [inline lib/scanctx.c]: dirname ? dirname : ctx->basedir

File: lib/scanctx.c [hunk 3]
```diff
@@ -146,13 +164,67 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
 
 /* ------------------------------------------------------------------------- */
 
+extern const char* scanctx_dirnext(struct scan_context* ctx)
+{
+  struct dirent** dentries= (struct dirent**)ctx->dentries;
+  if( NULL == ctx->dentries || ctx->de_cur==ctx->de_max )	/* shouldn't happen.... */
+    return NULL;
+
+  return dentries[ ctx->de_cur++ ]->d_name;
```
Comment (hyperrealm) [inline lib/scanctx.c]: no spaces after [ or before ]

File: lib/scanctx.c [hunk 4]
```diff
@@ -146,13 +164,67 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
+int scanctx_dirscan(struct scan_context* ctx, const char* dirname,
+                    int (*filter)(const struct dirent *),
+                    int (*compar)(const struct dirent **, const struct dirent **))
+{
+  int n;
+
+  if( NULL == dirname )return -1;
+
+  ctx->dentries=NULL;
+  if( (n=scandir(dirname,(struct dirent***)&ctx->dentries,filter,compar)) < 0)
```
Comment (hyperrealm) [inline lib/scanctx.c]: always put a space after ,

File: lib/scanctx.c (hunk 2 above)
Comment (hyperrealm) [inline lib/scanctx.c]: should be indented 2 spaces, not 1

File: lib/scanctx.c [hunk 5]
```diff
@@ -146,13 +164,67 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
 
   return(buffer);
 }
 
 /* ------------------------------------------------------------------------- */
 
+extern const char* scanctx_dirnext(struct scan_context* ctx)
+{
+  struct dirent** dentries= (struct dirent**)ctx->dentries;
+  if( NULL == ctx->dentries || ctx->de_cur==ctx->de_max )	/* shouldn't happen.... */
```
Comment (hyperrealm) [inline lib/scanctx.c]: but subexpressions in parentheses

File: lib/scanctx.c [hunk 6]
```diff
@@ -146,13 +164,70 @@ void *scanctx_pop_include(struct scan_context *ctx)
 
   --(ctx->depth);
   buffer = ctx->buffers[ctx->depth];
-  fclose(ctx->streams[ctx->depth]);
+  if(NULL!=ctx->streams[ctx->depth])
```
Comment (hyperrealm) [inline lib/scanctx.c]: drop the NULL!=

File: lib/scanctx.c [hunk 7]
```diff
@@ -146,13 +164,70 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
+                    int (*filter)(const struct dirent *),
+                    int (*compar)(const struct dirent **, const struct dirent **))
+{
+  int n;
+
+  if(!dirname)
+    return -1;
+
+  ctx->dentries = NULL;
+  if( (n=scandir(dirname, (struct dirent***)&ctx->dentries, filter, compar)) < 0)
```
Comment (hyperrealm) [inline lib/scanctx.c]: spacing: if((n = scandir...

File: lib/scanctx.c [hunk 8]
```diff
@@ -146,13 +164,70 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
+  ctx->dentries = NULL;
+
+  if(ctx->basedir)
+    free((void*)ctx->basedir);
+
+  ctx->de_cur = ctx->de_max = 0;
+
+  return 0;
+}
+int scanctx_inloop(const struct scan_context* ctx)
```
Comment (hyperrealm) [inline lib/scanctx.c]: add a blank line between functions

File: lib/scanctx.c [hunk 9]
```diff
@@ -146,13 +164,70 @@ void *scanctx_pop_include(struct scan_context *ctx)
...
+{
+  if(!ctx->dentries)
+    return 0;
+
+  return (ctx->de_cur < ctx->de_max)? 1 : 0;
```
Comment (hyperrealm) [inline lib/scanctx.c]: add space before ?

File: lib/scanctx.h [hunk 10]
```diff
@@ -38,18 +38,33 @@ struct scan_context
   const char *files[MAX_INCLUDE_DEPTH];
   void *buffers[MAX_INCLUDE_DEPTH];
   FILE *streams[MAX_INCLUDE_DEPTH];
-  int depth;
   strbuf_t string;
   const char **filenames;
   unsigned int num_filenames;
+  int depth;
+  void** dentries;		/* dirent** */
```
Comment (hyperrealm) [inline lib/scanctx.h]: void **dentries
const char *basedir

File: lib/scanner.l [hunk 11]
```diff
@@ -83,6 +84,22 @@ static unsigned long long fromhex(const char *s)
...
+{
+  const char *fname = de->d_name;
+
+#ifdef _DIRENT_HAVE_D_TYPE
+  /* filter out non-files and non-symlinks */
+  if((de->d_type != DT_REG) && (de->d_type != DT_LNK) && (de->d_type != DT_UNKNOWN))
+    return 0;
+#endif
+
+  return ( (fname != NULL)
```
Comment (hyperrealm) [inline lib/scanner.l]: return((fname != NULL)

File: lib/scanner.l [hunk 12]
```diff
@@ -126,33 +144,80 @@ include_open      ^[ \t]*@include[ \t]+\"
...
+
+        basedir = scanctx_getpath(yyextra);
+        if(scanctx_dirscan(yyextra, basedir, filter_dotfiles, alphasort) < 0)
+        {
+          if(basedir)
+            free((void*)basedir);
+          return TOK_ERROR;
+        }
+
+        if( scanctx_inloop(yyextra) )
```
Comment (hyperrealm) [inline lib/scanner.l]: remove the spaces within the parens

File: lib/scanner.l [hunk 13]
```diff
@@ -194,13 +259,40 @@ include_open      ^[ \t]*@include[ \t]+\"
...
-                    if(buf)
-                    {
-                      yy_delete_buffer(YY_CURRENT_BUFFER, yyscanner);
-                      yy_switch_to_buffer(buf, yyscanner);
-                    }
-                    else
-                      yyterminate();
-                  }
+        const char* error;
+        FILE* fp;
```
Comment (hyperrealm) [inline lib/scanner.l]: FILE *fp

**Pull request merged**